test/rule_test_support.dart

```diff
@@ -49,12 +61,164 @@ class AnalysisOptionsFileConfig {
   }
 }
 
+/// A description of a diagnostic that is expected to be reported.
+class ExpectedDiagnostic {
+  /// An empty array of diagnostic descriptors used when no errors are expected.
+  static List<ExpectedDiagnostic> noDiagnostics = <ExpectedDiagnostic>[];
+
+  final DiagnosticMatcher diagnosticMatcher;
```

Comment:
Personally I would prefer a style in which the subclasses override `matches` to invoke `super.matches` and then do the additional checks directly.